In [1]:
import QUANTAXIS as QA

QUANTAXIS>> start QUANTAXIS
QUANTAXIS>> Welcome to QUANTAXIS, the Version is 1.1.4
QUANTAXIS>>  
 ```````````````````````````````````````````````````````````````````````````````````````````````````````````````````````` 
  ``########`````##````````##``````````##`````````####````````##```##########````````#``````##``````###```##`````######`` 
  `##``````## ```##````````##`````````####````````##`##```````##```````##```````````###``````##````##`````##```##`````##` 
  ##````````##```##````````##````````##`##````````##``##``````##```````##``````````####```````#```##``````##```##``````## 
  ##````````##```##````````##```````##```##```````##```##`````##```````##`````````##`##```````##`##```````##````##``````` 
  ##````````##```##````````##``````##`````##``````##````##````##```````##````````##``###```````###````````##`````##`````` 
  ##````````##```##````````##``````##``````##`````##`````##```##```````##```````##````##```````###````````##``````###```` 
  ##````````##```##````````##`````##``````

# 在这里我们演示一下 下单/交易/结算的整个流程

我们首先会建立一个账户类和一个回测类


In [2]:
 # 初始化一个account
Account=QA.QA_Account()

# 初始化一个回测类
B = QA.QA_BacktestBroker()


在第一天的时候,全仓买入 000001

In [3]:
# 全仓买入'000001'

Order=Account.send_order(code='000001',
                        price=11,
                        money=0.04*Account.cash_available,
                        time='2018-05-09',
                        towards=QA.ORDER_DIRECTION.BUY,
                        order_model=QA.ORDER_MODEL.MARKET,
                        amount_model=QA.AMOUNT_MODEL.BY_MONEY
                        )




In [4]:

## 打印order的占用资金
print('ORDER的占用资金: {}'.format((Order.amount*Order.price)*(1+Account.commission_coeff)))

ORDER的占用资金: 39609.9


In [5]:
# 账户剩余资金
print('账户剩余资金 :{}'.format(Account.cash_available))

账户剩余资金 :960000.0


In [6]:
Account.hold

Series([], Name: amount, dtype: float64)

In [7]:
Account.init_hold

Series([], Name: amount, dtype: float64)

In [8]:
Account.hold_available

Series([], Name: amount, dtype: float64)

此时的账户cash并未减少,因为此过程为申报订单(已委托 未成交状态)

回测类接受订单 并返回订单(如果创建成功)

In [9]:
rec_mes=B.receive_order(QA.QA_Event(order=Order))
print(rec_mes)

< QA_Order realorder_id Order_dL4nhik6 datetime:2018-05-09 09:30:00 code:000001 amount:3600 price:10.955 towards:1 btype:stock_cn order_id:Order_dL4nhik6 account:Acc_uME9Gtp4 status:queued >


In [10]:
#B.query_orders(rec_mes.account_cookie,rec_mes.realorder_id)

In [11]:
import pandas as pd

In [12]:
B.query_orders(Account.account_cookie)

order_time    code  name  towards  \
account_cookie realorder_id                                                 
Acc_uME9Gtp4   Order_dL4nhik6  2018-05-09 09:31:00  000001  None        1   

                               trade_price  order_price         status  \
account_cookie realorder_id                                              
Acc_uME9Gtp4   Order_dL4nhik6        10.96       10.955  trade_success   

                               order_amount  trade_amount  cancel_amount  
account_cookie realorder_id                                               
Acc_uME9Gtp4   Order_dL4nhik6          3600          3600              0

In [13]:
#pd.DataFrame(list(B.deal_message.values()),columns=B.orderstatus_headers).set_index(['account_cookie','realorder_id'])

In [14]:
B.query_orders(Account.account_cookie,'filled')

trade_time    code  name  towards  \
account_cookie realorder_id                                                 
Acc_uME9Gtp4   Order_dL4nhik6  2018-05-09 00:00:00  000001  None        1   

                               trade_amount  trade_price  trade_money  \
account_cookie realorder_id                                             
Acc_uME9Gtp4   Order_dL4nhik6          3600        10.96   39447.8595   

                                     trade_id  
account_cookie realorder_id                    
Acc_uME9Gtp4   Order_dL4nhik6  Trade_4qeE9kXv

In [15]:
trade_mes=B.query_orders(Account.account_cookie,'filled')
res=trade_mes.loc[Order.account_cookie,Order.realorder_id]
Order.trade(res.trade_id,res.trade_price,res.trade_amount,res.trade_time)

receive deal


In [16]:
#Order.trade((trade_id, trade_price, trade_amount, trade_time))

账户类接收到回测返回的回报信息,更新账户

In [17]:
Account.history_table

datetime    code  price  amount        cash        order_id  \
0  2018-05-09 00:00:00  000001  10.96    3600  960534.136  Order_dL4nhik6   

     realorder_id        trade_id account_cookie  commission  tax  
0  Order_dL4nhik6  Trade_4qeE9kXv   Acc_uME9Gtp4       9.864    0

此时我们可以打印一下现在的状态(现在的状态可以理解为在交易时 买入一只000001股票,但当天尚未收盘)

In [18]:
print('账户的可用资金 {}'.format(Account.cash_available))

账户的可用资金 960534.1359999999


In [19]:
Account.hold

code
000001    3600
Name: amount, dtype: int64

In [20]:
Account.hold_available

code
000001    3600
Name: amount, dtype: int64

In [21]:
Account.init_hold.index_name='code'

In [22]:
import pandas as pd

In [23]:
pd.concat([Account.hold_available,Account.init_hold])

code
000001    3600
Name: amount, dtype: int64

In [24]:
+Account.hold_available

code
000001    3600
Name: amount, dtype: int64

我们注意到 当最初申报订单的时候,可用资金只有950.2999999999302元,而买入成功后,可用资金有3339.9289999998837元,原因是下单的时候模式是市价单模式(QA.ORDER_MODEL.MARKET),故实际成交金额为10.96元

买入以后 账户的持仓为90800股 000001

In [25]:
Account.hold

code
000001    3600
Name: amount, dtype: int64

买入后账户现金表被扩展

In [26]:
Account.cash

[1000000, 960534.1359999999]

因为是t+1的A股市场,故此时可卖数量为0

In [27]:
Account.sell_available

Series([], Name: amount, dtype: float64)

# 执行结算

In [28]:
Account.settle()

# 结算后

In [29]:
Account.cash

[1000000, 960534.1359999999]

In [30]:
Account.cash_available

960534.1359999999

In [31]:
Account.sell_available

code
000001    3600
Name: amount, dtype: int64

In [32]:
Account.hold

code
000001    3600
Name: amount, dtype: int64

# 执行卖出操作

现在的持仓为: 000001 90800股

In [33]:
holdnum=Account.sell_available.get('000001',0)

In [34]:
holdnum

3600

申报一个卖出单,把可卖全部卖出

In [35]:
Order=Account.send_order(code='000001',
                        price=11,
                        amount=holdnum,
                        time='2018-05-10',
                        towards=QA.ORDER_DIRECTION.SELL,
                        order_model=QA.ORDER_MODEL.MARKET,
                        amount_model=QA.AMOUNT_MODEL.BY_AMOUNT
                        )


In [36]:
Order

< QA_Order realorder_id Order_xaQklGso datetime:2018-05-10 09:31:00 code:000001 amount:3600 price:11 towards:-1 btype:stock_cn order_id:Order_xaQklGso account:Acc_uME9Gtp4 status:queued >

In [37]:
Account.cash_available  # 因为此时订单尚未申报成功 可用现金不变

960534.1359999999

In [38]:
rec_mes=B.receive_order(QA.QA_Event(order=Order))
print(rec_mes)

< QA_Order realorder_id Order_xaQklGso datetime:2018-05-10 09:30:00 code:000001 amount:3600 price:11.0 towards:-1 btype:stock_cn order_id:Order_xaQklGso account:Acc_uME9Gtp4 status:queued >


In [39]:
trade_mes=B.query_orders(Account.account_cookie,'filled')
res=trade_mes.loc[Order.account_cookie,Order.realorder_id]
Order.trade(res.trade_id,res.trade_price,res.trade_amount,res.trade_time)

receive deal


In [40]:
Account.cash_available # 此时订单已成交 cash_available立刻结转

1000084.6359999999

In [41]:
Account.history_table

datetime    code  price  amount         cash        order_id  \
0  2018-05-09 00:00:00  000001  10.96    3600   960534.136  Order_dL4nhik6   
1  2018-05-10 00:00:00  000001  11.00   -3600  1000084.636  Order_xaQklGso   

     realorder_id        trade_id account_cookie  commission   tax  
0  Order_dL4nhik6  Trade_4qeE9kXv   Acc_uME9Gtp4       9.864   0.0  
1  Order_xaQklGso  Trade_1ckWO95u   Acc_uME9Gtp4       9.900  39.6

In [42]:
Account.orders

<QA_ORDERQueue>

In [43]:
Account.orders.order_list

{'Order_dL4nhik6': < QA_Order realorder_id Order_dL4nhik6 datetime:2018-05-09 09:30:00 code:000001 amount:3600 price:10.955 towards:1 btype:stock_cn order_id:Order_dL4nhik6 account:Acc_uME9Gtp4 status:success_all >,
 'Order_xaQklGso': < QA_Order realorder_id Order_xaQklGso datetime:2018-05-10 09:30:00 code:000001 amount:3600 price:11.0 towards:-1 btype:stock_cn order_id:Order_xaQklGso account:Acc_uME9Gtp4 status:success_all >}

# 测试T0账户

In [44]:
 # 初始化一个account
AccountT0=QA.QA_Account(running_environment=QA.RUNNING_ENVIRONMENT.TZERO,init_hold={'000001':10000},init_cash=200000)

# 初始化一个回测类
B = QA.QA_BacktestBroker()

In [45]:
AccountT0.init_assets

{'cash': 200000, 'hold': {'000001': 10000}}

In [46]:
AccountT0.init_hold

code
000001    10000
Name: amount, dtype: int64

In [47]:
AccountT0.hold_available

Series([], Name: amount, dtype: float64)

In [48]:
AccountT0.sell_available

code
000001    10000
Name: amount, dtype: int64

In [49]:
Order=AccountT0.send_order(code='000001',
                        price=11,
                        amount=AccountT0.sell_available.get('000001',0),
                        time='2018-05-10',
                        towards=QA.ORDER_DIRECTION.SELL,
                        order_model=QA.ORDER_MODEL.MARKET,
                        amount_model=QA.AMOUNT_MODEL.BY_AMOUNT
                        )


In [50]:
Order.datetime

'2018-05-10 09:31:00'

In [51]:
rec_mes=B.receive_order(QA.QA_Event(order=Order))
print(rec_mes)

< QA_Order realorder_id Order_aFpZ98Yk datetime:2018-05-10 09:30:00 code:000001 amount:10000 price:11.0 towards:-1 btype:stock_cn order_id:Order_aFpZ98Yk account:Acc_WTAuvjiX status:queued >


In [52]:
rec_mes

< QA_Order realorder_id Order_aFpZ98Yk datetime:2018-05-10 09:30:00 code:000001 amount:10000 price:11.0 towards:-1 btype:stock_cn order_id:Order_aFpZ98Yk account:Acc_WTAuvjiX status:queued >

In [53]:
B.query_orders(AccountT0.account_cookie,'filled')

trade_time    code  name  towards  \
account_cookie realorder_id                                                 
Acc_WTAuvjiX   Order_aFpZ98Yk  2018-05-10 00:00:00  000001  None       -1   

                               trade_amount  trade_price  trade_money  \
account_cookie realorder_id                                             
Acc_WTAuvjiX   Order_aFpZ98Yk         10000         11.0     110137.5   

                                     trade_id  
account_cookie realorder_id                    
Acc_WTAuvjiX   Order_aFpZ98Yk  Trade_soEa4gQc

In [54]:
trade_mes=B.query_orders(AccountT0.account_cookie,'filled')
res=trade_mes.loc[Order.account_cookie,Order.realorder_id]
Order.trade(res.trade_id,res.trade_price,res.trade_amount,res.trade_time)

receive deal


In [55]:
AccountT0.sell_available

code
000001    0
Name: amount, dtype: int64

In [56]:
Order.trade_time

['2018-05-10 00:00:00']

In [57]:
AccountT0.buy_available

code
000001    10000
Name: amount, dtype: int64

In [58]:
AccountT0.hold_available

code
000001   -10000
Name: amount, dtype: int64

In [59]:
AccountT0.running_time

datetime.datetime(2018, 9, 5, 11, 41, 53, 764872)

In [60]:
AccountT0.datetime

'2018-05-10 00:00:00'

In [61]:
AccountT0.buy_available.get('000001')

10000

In [62]:
r=AccountT0.close_positions_order

In [63]:
AccountT0.date

'2018-05-10'

In [64]:
for Order in r:
    print(vars(Order))
    rec_mes=B.receive_order(QA.QA_Event(order=Order))
    trade_mes=B.query_orders(AccountT0.account_cookie,'filled')
    res=trade_mes.loc[Order.account_cookie,Order.realorder_id]
    Order.trade(res.trade_id,res.trade_price,res.trade_amount,res.trade_time)

{'price': 0, 'datetime': '2018-05-10 15:00:00', 'date': '2018-05-10', 'sending_time': '2018-05-10 15:00:00', 'trade_time': [], 'amount': 10000, 'trade_amount': 0, 'cancel_amount': 0, 'towards': 1, 'code': '000001', 'user': None, 'market_type': 'stock_cn', 'frequence': 'day', 'account_cookie': 'Acc_WTAuvjiX', 'strategy': None, 'type': 'stock_cn', 'order_model': 'CLOSE', 'amount_model': 'by_amount', 'order_id': 'Order_rgq2nGku', 'realorder_id': 'Order_rgq2nGku', 'commission_coeff': 0.00025, 'tax_coeff': 0.001, 'trade_id': [], 'trade_price': 0, 'callback': <bound method QA_Account.receive_deal of < QA_Account Acc_WTAuvjiX>>, 'money': 0.0, 'reason': None, '_status': 'new'}
receive deal


In [65]:
AccountT0.cash

[200000, 309862.5, 199734.975]

In [66]:
AccountT0.sell_available

code
000001    0
Name: amount, dtype: int64

In [96]:
AccountT0.hold_available

Series([], Name: amount, dtype: float64)

In [97]:
AccountT0.hold

code
000001    10000
Name: amount, dtype: int64

In [98]:
AccountT0.settle()

In [99]:
AccountT0.daily_hold

code                       000001
date       account_cookie        
2018-05-10 Acc_WTAuvjiX         0

In [100]:
AccountT0.daily_cash

cash             datetime       date  \
date       account_cookie                                           
2018-05-10 Acc_WTAuvjiX    199735  2018-05-10 00:00:00 2018-05-10   

                          account_cookie  
date       account_cookie                 
2018-05-10 Acc_WTAuvjiX     Acc_WTAuvjiX

In [72]:
AccountT0.hold_table()

code
000001    10000
Name: amount, dtype: int64

In [73]:
AccountT0.hold_price()

Series([], dtype: float64)

In [74]:
AccountT0.datetime

'2018-05-11 09:30:00'

In [75]:
AccountT0.sell_available

code
000001    10000
Name: amount, dtype: int64

In [76]:
risk_t0=QA.QA_Risk(AccountT0)

In [77]:
risk_t0.init_assets

{'cash': 200000, 'hold': {'000001': 10000}}

In [78]:
risk_t0.init_cash-risk_t0.assets

date
2018-05-10    265.025
Name: 0, dtype: float64

In [79]:
risk_t0.assets.iloc[-1]/risk_t0.init_cash-1

-0.001325124999999927

In [80]:
risk_t0.profit_construct

{'total_buyandsell': -99.99,
 'total_tax': -110.0,
 'total_commission': -55.03,
 'total_profit': -265.02}

# 测试期货账户

In [81]:
rb=QA.QA_fetch_get_future_min('tdx','RBL8','2018-01-01','2018-02-28',frequence='15min')
rb_ds=QA.QA_DataStruct_Future_min(rb.set_index(['datetime','code']))

AccountFuture=QA.QA_Account(init_cash=1000000,allow_sellopen=True,allow_t0=True,account_cookie='future_test',market_type=QA.MARKET_TYPE.FUTURE_CN,frequence=QA.FREQUENCE.FIFTEEN_MIN)


#Account.reset_assets(10000000)
Broker=QA.QA_BacktestBroker()


QUANTAXIS>> Selecting the Best Server IP of TDX


USING DEFAULT STOCK IP
USING DEFAULT FUTURE IP


QUANTAXIS>> === The BEST SERVER ===
 stock_ip 60.191.117.167 future_ip 120.24.0.77


In [82]:
rb

open    high     low   close  position   trade  price  \
datetime                                                                       
2018-01-02 09:15:00  3801.0  3809.0  3794.0  3798.0   2520510  312684    0.0   
2018-01-02 09:30:00  3799.0  3834.0  3797.0  3830.0   2501350  418536    0.0   
2018-01-02 09:45:00  3829.0  3831.0  3819.0  3822.0   2498648  137928    0.0   
2018-01-02 10:00:00  3823.0  3824.0  3812.0  3822.0   2501706  116806    0.0   
2018-01-02 10:15:00  3822.0  3825.0  3818.0  3824.0   2510106   68888    0.0   
2018-01-02 10:45:00  3824.0  3828.0  3821.0  3823.0   2513434   71846    0.0   
2018-01-02 11:00:00  3824.0  3833.0  3820.0  3824.0   2526064  156954    0.0   
2018-01-02 11:15:00  3824.0  3830.0  3821.0  3826.0   2521712   77962    0.0   
2018-01-02 11:30:00  3825.0  3872.0  3819.0  3871.0   2464210  538316    0.0   
2018-01-02 13:45:00  3872.0  3879.0  3867.0  3871.0   2437082  339672    0.0   
2018-01-02 14:00:00  3873.0  3874.0  3862.0  3866.0   2441464  122992    0.0   
2018-01-02 14:15:00  3866.0  3875.0  3864.0  3875.0   2448522   93220    0.0   
2018-01-02 14:30:00  3874.0  3877.0  3867.0  3873.0   2455406   89550    0.0   
2018-01-02 14:45:00  3873.0  3886.0  3872.0  3879.0   2449654  209192    0.0   
2018-01-02 15:00:00  3878.0  3880.0  3869.0  3874.0   2413572  201602    0.0   
2018-01-03 21:15:00  3866.0  3866.0  3855.0  3863.0   2440272  253022    0.0   
2018-01-03 21:30:00  3864.0  3868.0  3857.0  3860.0   2461428  124270    0.0   
2018-01-03 21:45:00  3861.0  3873.0  3858.0  3871.0   2461948  128146    0.0   
2018-01-03 22:00:00  3871.0  3872.0  3858.0  3864.0   2472196  100256    0.0   
2018-01-03 22:15:00  3864.0  3873.0  3857.0  3867.0   2472810  122358    0.0   
2018-01-03 22:30:00  3867.0  3872.0  3865.0  3868.0   2475272   59818    0.0   
2018-01-03 22:45:00  3869.0  3871.0  3860.0  3868.0   2471526   68318    0.0   
2018-01-03 23:00:00  3868.0  3869.0  3860.0  3863.0   2449096   71270    0.0   
2018-01-03 09:15:00  3863.0  3866.0  3845.0  3854.0   2445824  250624    0.0   
2018-01-03 09:30:00  3853.0  3858.0  3851.0  3857.0   2452682   88408    0.0   
2018-01-03 09:45:00  3856.0  3859.0  3852.0  3855.0   2459606   59256    0.0   
2018-01-03 10:00:00  3855.0  3876.0  3854.0  3873.0   2454690  157042    0.0   
2018-01-03 10:15:00  3874.0  3875.0  3866.0  3866.0   2462512   89868    0.0   
2018-01-03 10:45:00  3866.0  3868.0  3852.0  3856.0   2469486  120010    0.0   
2018-01-03 11:00:00  3857.0  3859.0  3854.0  3857.0   2472910   43056    0.0   
...                     ...     ...     ...     ...       ...     ...    ...   
2018-02-27 11:30:00  4002.0  4006.0  3982.0  3992.0   2499834  272200    0.0   
2018-02-27 13:45:00  3993.0  4009.0  3992.0  4007.0   2503190  134484    0.0   
2018-02-27 14:00:00  4007.0  4015.0  4004.0  4014.0   2501602  120106    0.0   
2018-02-27 14:15:00  4014.0  4026.0  4014.0  4024.0   2518878  194990    0.0   
2018-02-27 14:30:00  4024.0  4027.0  4018.0  4025.0   2515318   77490    0.0   
2018-02-27 14:45:00  4025.0  4038.0  4022.0  4037.0   2529884  206128    0.0   
2018-02-27 15:00:00  4036.0  4040.0  4029.0  4035.0   2509598  180976    0.0   
2018-02-28 21:15:00  4030.0  4030.0  4011.0  4014.0   2500524  200510    0.0   
2018-02-28 21:30:00  4015.0  4018.0  4007.0  4016.0   2483946  126236    0.0   
2018-02-28 21:45:00  4017.0  4024.0  4016.0  4021.0   2481956   63478    0.0   
2018-02-28 22:00:00  4021.0  4022.0  4018.0  4018.0   2487526   29260    0.0   
2018-02-28 22:15:00  4018.0  4031.0  4018.0  4031.0   2499250   90886    0.0   
2018-02-28 22:30:00  4030.0  4031.0  4021.0  4028.0   2501550   73888    0.0   
2018-02-28 22:45:00  4028.0  4034.0  4022.0  4031.0   2518590   74228    0.0   
2018-02-28 23:00:00  4031.0  4032.0  4024.0  4025.0   2501768   58694    0.0   
2018-02-28 09:15:00  4023.0  4024.0  4009.0  4014.0   2491540  173826    0.0   
2018-02-28 09:30:00  4014.0  4017.0  4011.0  4012.0   2482078   60094    0.0   
2018-02-28 09:45:00  40

In [83]:
# 全仓买入'000001'
hq=next(rb_ds.panel_gen)
order=AccountFuture.send_order(code='RBL8',
                        price=hq.open,
                        money=AccountFuture.cash_available,
                        time=hq.datetime[0],
                        towards=QA.ORDER_DIRECTION.BUY,
                        order_model=QA.ORDER_MODEL.MARKET,
                        amount_model=QA.AMOUNT_MODEL.BY_MONEY
                        )


In [84]:
Broker.receive_order(QA.QA_Event(order=order,market_data=hq))

< QA_Order realorder_id Order_1YzjQXi6 datetime:2018-01-02 09:15:00 code:RBL8 amount:200 price:3801.5 towards:1 btype:future_cn order_id:Order_1YzjQXi6 account:future_test status:queued >

In [85]:
trade_mes = Broker.query_orders(AccountFuture.account_cookie, 'filled')
res = trade_mes.loc[order.account_cookie, order.realorder_id]
order.trade(res.trade_id, res.trade_price, res.trade_amount, res.trade_time)

receive deal
